packages/block-editor/src/components/content-only-controls/index.js

```diff
@@ -80,40 +156,201 @@ function BlockFields( { clientId } ) {
 		[ clientId ]
 	);
 
+	const { updateBlockAttributes } = useDispatch( blockEditorStore );
 	const blockTitle = useBlockDisplayTitle( {
 		clientId,
 		context: 'list-view',
 	} );
 	const blockInformation = useBlockDisplayInformation( clientId );
-	const popoverPlacementProps = useInspectorPopoverPlacement();
 
-	if ( ! blockType?.[ fieldsKey ]?.length ) {
+	const blockTypeFields = blockType?.[ fieldsKey ];
+
+	// Track visible fields
+	const [ visibleFields, setVisibleFields ] = useState( () => {
+		// Show fields that have shownByDefault: true by default
+		return (
+			blockTypeFields
+				?.filter( ( field ) => field.shownByDefault )
+				.map( ( field ) => field.id ) || []
+		);
+	} );
+
+	// Build DataForm fields with proper structure
+	const dataFormFields = useMemo( () => {
+		if ( ! blockTypeFields?.length ) {
+			return [];
+		}
+
+		return blockTypeFields.map( ( fieldDef ) => {
+			const ControlComponent = CONTROLS[ fieldDef.type ];
+
+			const defaultValues = {};
+			if ( fieldDef.mapping && blockType?.attributes ) {
+				Object.entries( fieldDef.mapping ).forEach(
+					( [ key, attrKey ] ) => {
+						defaultValues[ key ] =
+							blockType.attributes[ attrKey ]?.defaultValue ??
+							undefined;
+					}
+				);
+			}
+
+			const field = {
+				id: fieldDef.id,
+				label: fieldDef.label,
+				type: fieldDef.type, // Use the field's type; DataForm will use built-in or custom Edit
+				config: { ...fieldDef.args, defaultValues },
```

Comment:
Not for now, but potentially in a follow-up we might move these from `fieldDef.args` to be properties of `fieldDef.Edit`. Similarly the config in _this_ object might move to `Edit` config, too. I haven't gone further in this PR because I wanted to kind of mark a point of iteration and see if this is good enough for now.